File: examples/java/gmpub_wsmq_async.java
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** 
 * @file gmpub_wsmq_async.java
 *
 * An example Java GMSEC publisher.
 */

import gov.nasa.gsfc.gmsec.api.*;
import gov.nasa.gsfc.gmsec.api.util.Log;


public class gmpub_wsmq_async implements Example
{
	private Config     config;
	private Connection connection;

	// This callback is user-defined and can be used to handle the Status object returned by this event
	class AsyncStatusCheckCallback extends EventCallback
	{
		public void onEvent(Connection conn, Status status, Connection.ConnectionEvent event)
		{
			Log.info(status.getReason());

			if (status.isError())
			{
				Log.error("The first occurrence of a WebSphere MQ Asynchronous Put Response warning or failure returned the WebSphere Reason Code: " + status.getCustomCode());
			}
		}
	}

	public gmpub_wsmq_async(Config config) throws GMSEC_Exception, ExampleException
	{
		this.config = config;

		Util.initialize(config);

		if (!Util.hasConnectionType(config))
		{
			printUsage();
			throw new ExampleException("Invalid option");
		}
	}


	private void printUsage()
	{
		System.err.println("\nusage: java gmpub_wsmq_async connectionType=<middleware> " +
		          "[ <parameter>=<value> ]\n" +
		          "\n\tNote that the parameter 'connectiontype' is required. "  +
		          "\n\tThe rest of other parameters are optional.\n" +
		          "\n\tserver=<server name> " +
		          "(required if middleware is not bolt/MB locally)" +
		          "\n\tsubject=<subject name>" +
		          "\n\titerations=<iterations>" +
		          "\n\tinterval_ms=<interval (milliseconds)>" +
		          "\n\tloglevel=<log level>" +
		          "\n\tcfgfile=<config_filepath.xml>" +
	                  "\n\tmw-async-publish=<true or false>" +
		          "\n\tmw-async-status-check=<true or false>" +
		          "\n\tmw-async-status-check-message-interval=<number of publish operations before checking asynchronous put status>" +
		          "\n\nFor more information, see API User's Guide\n\n");
	}


	public boolean run() throws GMSEC_Exception
	{
		// Output GMSEC API version
		Log.info(Connection.getAPIVersion());

		String subject     = Util.get(config, "SUBJECT", "GMSEC.TEST.PUBLISH");
		int    iterations  = Util.get(config, "ITERATIONS", 1);
		int    interval_ms = Util.get(config, "INTERVAL_MS", 1000);

		connection = Util.openConnection(config);

		// Output middleware version
		Log.info("Middleware version = " + connection.getLibraryVersion());

		/// Set up the WebSphere Asynchronous Publish Status Checker Event Callback
		connection.registerEventCallback(Connection.ConnectionEvent.WSMQ_ASYNC_STATUS_CHECK_EVENT, new AsyncStatusCheckCallback());

		// Publish messages
		for (int i = 0; i < iterations; ++i)
		{
			publish(i, subject);

			if (i < iterations - 1)
			{
				Util.millisleep(interval_ms);
			}
		}

		return true;
	}


	void publish(int iteration, String subject) throws GMSEC_Exception
	{
		int value = iteration + 1;

		Message message = new Message(subject, Message.MessageKind.PUBLISH);

		// Add Fields
		String data = "JLMNOPQ";

		message.addField("BIN-FIELD", data.getBytes());

		message.addField("BOOL-FIELD-TRUE", true);
		message.addField("BOOL-FIELD-FALSE", false);

		message.addField("CHAR-FIELD", 'c');

		message.addField("I8-FIELD", (byte) value);
		message.addField("I16-FIELD", (short) value);
		message.addField("I32-FIELD", value);
		message.addField("I64-FIELD", (long) value);

		message.addField("U8-FIELD", new U8(value % (U8.MAX_VALUE + 1)));
		message.addField("U16-FIELD", new U16(value % (U16.MAX_VALUE + 1)));
		message.addField("U32-FIELD", new U32(value % (U32.MAX_VALUE + 1)));
		message.addField("U64-FIELD", new U64(Integer.toString(value)));

		message.addField("STRING-FIELD", "This is a test");

		message.addField("F32-FIELD", (float)(1 + 1. / value));
		message.addField("F64-FIELD", (double)(1 + 1. / value));

		message.addField("COUNT", iteration);


		String encrypt = config.getValue("ENCRYPT");
		if (encrypt != null) {
			message.addField("SEC-ENCRYPT", true);
		}

		// Publish Message
		connection.publish(message);

		// Display XML representation of the message
		Log.info("Published:\n" + message.toXML());
	}


	public boolean cleanup() throws GMSEC_Exception
	{
		if (connection != null)
		{
			Util.closeConnection(connection);
		}

		return true;
	}


	public static void main(String[] args)
	{
		Example example = null;

		try {
			Config config = new Config(args);

			Util.addToConfigFromFile(config);

			example = new gmpub_wsmq_async(config);
			example.run();
		}
		catch (IllegalArgumentException e) {
			System.err.println(e.toString());
		}
		catch (GMSEC_Exception e) {
			System.err.println(e.toString());
		}
		catch (ExampleException e) {
			System.err.println(e.toString());
		}
		finally {
			if (example != null)
				Util.cleanup(example);
		}
	}
}

File: examples/legacy/cpp/gmpub_wsmq_async/gmpub_wsmq_async.cpp
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** 
 * @file gmpub_wsmq_async.cpp
 * 
 * A C++ example demonstration of GMSEC publish/subscribe functionality.
 * The associated example gmsub.cpp will listen for this publish program.
 * gmsub should be run before gmpub_wsmq_async.
 * 
 */

#include "../example.h"

#include <string>

using namespace gmsec::api;
using namespace gmsec::api::util;


// This callback is user-defined and can be used to handle the Status object returned by this event
class AsyncStatusCheckCallback : public EventCallback
{
public:
	void CALL_TYPE onEvent(Connection& conn, const Status& status, Connection::ConnectionEvent event)
	{
		GMSEC_INFO << status.getReason();

		if (status.isError())
		{
			GMSEC_ERROR << "The first occurrence of a WebSphere MQ Asynchronous Put Response warning or failure returned the WebSphere Reason Code: " << status.getCustomCode();
		}
	}
};


class gmpub_wsmq_async
{
public:
	gmpub_wsmq_async(Config& config);
	~gmpub_wsmq_async();
	bool run();

private:
	Config&     config;
	Connection* connection;

	void publish(int iteration, const std::string& subject);
};


gmpub_wsmq_async::gmpub_wsmq_async(Config &c)
	: config(c),
	  connection(0)
{
	/* Initialize config */
	example::initialize(c);
}


gmpub_wsmq_async::~gmpub_wsmq_async()
{
	if (connection)
	{
		connection->disconnect();

		Connection::destroy(connection);
	}

	Connection::shutdownAllMiddlewares();
}


bool gmpub_wsmq_async::run()
{
	AsyncStatusCheckCallback asccb;	
	bool success = true;

	/* output GMSEC API version */
	GMSEC_INFO << Connection::getAPIVersion();

	try
	{
		//o Create the Connection
		connection = Connection::create(config);

		//o Connect
		connection->connect();

		//o Register the event callback with the connection to catch the WebSphere asynchronous publish status events
		connection->registerEventCallback(Connection::WSMQ_ASYNC_STATUS_CHECK_EVENT, &asccb);

		//o Get information from the command line
		std::string subject     = example::get(config, "SUBJECT", "GMSEC.TEST.PUBLISH");
		int         iterations  = example::get(config, "ITERATIONS", 1);
		int         interval_ms = example::get(config, "INTERVAL_MS", 1000);

		//o Output middleware version
		GMSEC_INFO << "Middleware version = " << connection->getLibraryVersion();

		//o Output information
		GMSEC_INFO << "Using subject '" << subject.c_str() << "'";
		GMSEC_INFO << "Publishing " << iterations << " message(s)";

		if (iterations > 1)
		{
			GMSEC_INFO << "Publish interval " << interval_ms << " [ms]";
		}

		//o Publish message(s)
		for (int i = 0; i < iterations; ++i)
		{
			publish(i, subject);

			if (i < iterations - 1)
			{
				gmsec::api::example::millisleep(interval_ms);
			}
		}
	}
	catch (Exception& e)
	{
		GMSEC_ERROR << e.what();
		success = false;
	}

	return success;
}


void gmpub_wsmq_async::publish(int iteration, const std::string& subject)
{
	int value = iteration + 1;

	GMSEC_INFO << "working";

	//o Create message
	Message message(subject.c_str(), Message::PUBLISH);

	//o Add fields to message
	message.addField("CHAR-FIELD", (GMSEC_CHAR) 'c');
	message.addField("BOOL-FIELD-TRUE", true);
	message.addField("BOOL-FIELD-FALSE", false);
	message.addField("I8-FIELD", (GMSEC_I8) value);
	message.addField("I16-FIELD", (GMSEC_I16) value);
	message.addField("I32-FIELD", (GMSEC_I32) value);
	message.addField("I64-FIELD", (GMSEC_I64) value);
	message.addField("U8-FIELD", (GMSEC_U8) value);
	message.addField("U16-FIELD", (GMSEC_U16) value);
	message.addField("U32-FIELD", (GMSEC_U32) value);
	message.addField("U64-FIELD", (GMSEC_U64) value);
	message.addField("COUNT", (GMSEC_I32) iteration);
	message.addField("STRING-FIELD", "This is a test");
	message.addField("F32-FIELD", GMSEC_F32(1 + 1. / value));
	message.addField("F64-FIELD", GMSEC_F64(1 + 1. / value));
	message.addField("BIN-FIELD", (GMSEC_BIN) "JLMNOPQ", 7);

	if (config.getBooleanValue("ENCRYPT", false))
	{
		message.addField("SEC-ENCRYPT", true);
	}

	//o Publish Message
	connection->publish(message);

	//o Display XML representation of message
	GMSEC_INFO << "Published:\n" << message.toXML();
}


int main(int argc, char* argv[])
{
	Config config(argc, argv);

	example::addToConfigFromFile(config);

	if (example::isOptionInvalid(config, argc))
	{
		example::printUsage("gmpub_wsmq_async");
		return -1;
	}

	gmpub_wsmq_async(config).run();
}
